### PR TITLE
Alliances

### PR DESCRIPTION
**Pull request by geoffthemedio**
Adds support for testing alliance status separate from peace in conditions / effects.

Adds alliance-related responses to AIs.

### CODE REVIEW

**Comment by o01eg** [comment]
> /freeorion/test/parse/TestEnumParser.cpp:125: error: fatal error: in "TestEnumParser/EmpireAffiliationTypeParser": Untested enumeration value.

File: default/python/AI/DiplomaticCorp.py [hunk 1]
```diff
@@ -36,6 +36,42 @@ def handle_diplomatic_message(self, message):
...
+                if attitude > 10:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_YES_MILD_LIST")
+                else:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_NO_MILD_LIST")
+            else:
+                possible_acknowledgments = UserStringList("AI_ALLIANCE_PROPOSAL_ACKNOWLEDGEMENTS_HARSH_LIST")
+                if attitude > 10:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_YES_HARSH_LIST")
+                else:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_NO_HARSH_LIST")
```
Comment (Morlic-fo) [inline default/python/AI/DiplomaticCorp.py:64]: The ```attitude``` calcs look out of place and seem error prone.

1) Current definition is that ```character.attitude_to_empire``` is limited to range [-10 .. 10]
2) The character already considers aggression trait when determining the empire's attitude.

Unless I am missing something, the current code is equivalent to the following (which I strongly prefer for clarity):
```
if aggression.key == fo.aggression.beginner:
    accept_proposal = True
elif aggression.key == fo.aggression.turtle:
    accept_proposal = attitude > 0
elif aggression.key == fo.aggression.cautious:
    accept_proposal = attitude > 5
else:  # aggression typical or greater
    accept_proposal = False
```

In particular, AIs with aggressiveness at least typical will never accept the alliance. That is not a gameplay problem considering the AI does not really handle allied status anyway, so preventing higher "difficulty" AI from doing something stupid is a good thing. However, that fact is not obvious from reading the proposed implementation.

**Review by Morlic-fo — CHANGES_REQUESTED** [review]

File: default/python/AI/DiplomaticCorp.py (hunk 1 above)
Comment (geoffthemedio) [reply default/python/AI/DiplomaticCorp.py:64]: hopefully implemented...

File: default/python/AI/DiplomaticCorp.py [hunk 2]
```diff
@@ -36,6 +36,42 @@ def handle_diplomatic_message(self, message):
         if message.recipient != fo.empireID():
             return
         aistate = get_aistate()
+        if message.type == fo.diplomaticMessageType.alliesProposal:
```
[the remark this answers is not in the document]
Comment (geoffthemedio) [reply default/python/AI/DiplomaticCorp.py:39]: done

File: universe/Condition.cpp [hunk 3]
```diff
@@ -1187,13 +1187,23 @@ namespace {
                 break;
             }
 
+            case AFFIL_PEACE: {
+                if (m_empire_id == ALL_EMPIRES)
+                    return true;
```
[the remark this answers is not in the document]
Comment (geoffthemedio) [reply universe/Condition.cpp]: fixed

**Review by Morlic-fo — APPROVED** [review]

**Comment by Morlic-fo** [comment]
AI changes look fine.

Please note that e304208 and 14d4668 address bugs which are currently present in master.

**Pull request merged**